gprofiler/profilers/node.py

```diff
@@ -0,0 +1,244 @@
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
+    # so it will always the same. When another debugger is opened in same NS it will not open new one.
+    # REF: Inspector agent initialization uses host_port
+    # https://github.com/nodejs/node/blob/5fad0b93667ffc6e4def52996b9529ac99b26319/src/inspector_agent.cc#L668
+    # host_port defaults to 9229
+    # ref: https://github.com/nodejs/node/blob/2849283c4cebbfbf523cc24303941dc36df9332f/src/node_options.h#L90
+    # in our case it won't be changed
+    port = 9229
+    debugger_url_response = requests.get(f"http://127.0.0.1:{port}/json/list")
+    if debugger_url_response.status_code != 200 or "application/json" not in debugger_url_response.headers.get(
+        "Content-Type", ""
+    ):
+        raise NodeDebuggerUrlNotFound(
+            {"status_code": debugger_url_response.status_code, "text": debugger_url_response.text}
+        )
+
+    response_json = debugger_url_response.json()
+    if (
+        not isinstance(response_json, list)
+        or len(response_json) == 0
+        or not isinstance(response_json[0], dict)
+        or "webSocketDebuggerUrl" not in response_json[0]
+    ):
+        raise NodeDebuggerUrlNotFound(response_json)
+
+    return cast(str, response_json[0]["webSocketDebuggerUrl"])
+
+
+def _send_socket_request(sock: WebSocket, cdp_request: Dict) -> None:
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+    if (
+        "result" not in message.keys()
+        or "result" not in message["result"].keys()
+        or "type" not in message["result"]["result"].keys()
+        or message["result"]["result"]["type"] != "boolean"
+    ):
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+def _get_node_pid(sock: WebSocket) -> int:
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": "process.pid",
+            "replMode": True,
+        },
+    }
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+    try:
+        return cast(int, message["result"]["result"]["value"])
+    except KeyError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+def _load_dso(sock: WebSocket, module_path: str) -> None:
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": f'process.mainModule.require("{os.path.join(module_path, "linux-perf.js")}").start()',
+            "replMode": True,
+        },
+    }
+    _send_socket_request(sock, cdp_request)
+
+
+def _stop_dso(sock: WebSocket, module_path: str) -> None:
```

Comment:
This function is exactly like `_load_dso`, only `start` became `stop`. Coalesce them?